### ISSUE TITLE
SearchGraph sliently moved

### ISSUE THREAD

**Issue by ben-strasser** [GVeitBatz/KaTCH#2]
Hi,

I needed quite some time to notice that the constructor TchEAQuery always moves the graph, even though no std::move is present at the caller site. More usually C++ patterns are either passing the graph by const reference or passing it by value. In the later case the caller can use std::move to elude the copy. The advantage of this pattern is that there is an indication at the caller site that the internals of the caller's graph object have been ripped out. This indication is missing in your code.

Another problem is that I need the graph to lookup details of the extracted path, such as for example the TTF of each edge. With the current design, this is not possible. The only graph instance has been moved into the query object. The query object has no get_graph function and therefore the graph has become inaccessible to the caller.

The cleanest solution would in my eyes be to have TchEAQuery only reference the graph object and not copy it. This way you could also have several query objects, as needed for multiple worker threads, on the same graph.

The easier to implement solution is to remove the && from the constructor, add std::move to the examples, and add a get_graph method.

Best Regards
Ben

**Comment by GVeitBatz** [comment]
Hi,
first of all I agree that the current design is not good. It would be much better to let the query store a pointer. An ok (but maybe not optimal) alternative would be to add a get_graph() function returning a const ref of the graph. This might serve as a simple fix.

> even though no std::move is present at the caller site.
But there is an std::move() at the caller site in the examples? Maybe I misunderstood what you want to point out.

>More usually C++ patterns are either passing the graph by const reference or passing it by value.
Well, with C++ eleven they invented these && references to indicate that moving is expectd by the interface. So, I think that is meant to be a common C++ pattern, even if maybe not fully popularized yet.

>In the later case the caller can use std::move to elude the copy. The advantage of this pattern is that there is an indication at the caller site that the internals of the caller's graph object have been ripped out.
You mean having a const ref or by-value formal parameter in the functions signature? I am quite convinced that this will not work. They way to do that is a non constant R-value reference plus R-value on caller site. std::move does nothing, it is only a type cast that makes a type to be an R-value. Or do you mean I should have both a move and a non-move version of the constructor? Right, that would be better.

But anyway, that is more a philosophical discussion. The design as it is is not good. I agree with that view.

>This indication is missing in your code.
As I said, isn't there std::move already there?

**Comment by ben-strasser** [comment]
> But there is an std::move() at the caller site in the examples?
> Maybe I misunderstood what you want to point out.

My mistake, I missed it.

> Well, with C++ eleven they invented these && references to indicate
> that moving is expectd by the interface. So, I think that is meant
> to be a common C++ pattern, even if maybe not fully popularized yet.

The && are mostly only needed in template code, move constructors, move 
assign and in some parts of the standard library, such as implementing 
std::move. auto&& is also an example. In terms of calling conventions 
the usual are:

void foo1(const MyType&x);

void foo2(MyType x);

void foo3(const MyType&x);
void foo3(MyType&&x);

foo1 never copies it's parameter but that comes at the price of the 
callee not being able to change x. The caller can use std::move but it 
has no effect.

With foo2 the callee always has his own object x that he can modify. The 
caller can decide where it comes from. The default is to create it by 
copy. The caller can turn this copy into a move using std::move.

With foo3 the callee can provide a different implementation depending on 
whether the caller is willing to give up control of his variable. The 
callee can decide using std::move, which variant is to be called. By 
convention both functions should do the same. Copy constructors and move 
constructors use this pattern.

foo1 and foo2 are definitely the most common. foo3 is rarely used 
outside of the move constructor pattern and is prone to subtle bugs as 
you need to have two code paths that are semantically nearly perfectly 
in sync at all times.

The following variant:

void foo4(MyType&&x);

forces that use to use std::move. It has to my knowledge only a niche 
usecase when in the context of constructors. You can use it to create a 
type that is only moveable.

Sidenote:

template<class T>
void foo5(T&&x);

has its own set of rules. Google for "universal reference" all of the 
details.

TLDR: In case of doubt, do not use && outside of move constructors and 
assignments.

Back to the actual topic:

In my local copy I added a get_graph function that returns a const&. 
That is definitely a work around that gets the job done. However, I'm 
uncertain whether the change to a pointer is that much more work. 
Replace all "_graph." with a "_graph->" in the right file should already 
be most of the work.

Best Regards
Ben
